### ISSUE TITLE
How to modularize and customise pages?

### ISSUE THREAD

**Issue by hroemer** [zauberzeug/nicegui#94]
The recent changes to page creation with the release of [0.9.0](https://github.com/zauberzeug/nicegui/releases/tag/v0.9.0) renders the example discussed in #50 unfeasible:

```python
class MainPage(ui.page):
    def __init__(self):
        super().__init__('/', dark=True)
        with self:
            ui.label('Hi!')

MainPage()
```

I have tried adapting the code (using PageBuilder) in the simple example, but failed to get it running as before.

```python
from nicegui import ui
from nicegui.globals import page_builders
from nicegui.page import Page
from nicegui.page_builder import PageBuilder


class MainPage(Page):
    def __init__(self):
        super().__init__(dark=True)

        ui.label('Hi!')


async def page_builder() -> Page:
    return MainPage()


def index_page():
    builder = PageBuilder(shared=False, function=page_builder)
    page_builders.setdefault('/', builder)


index_page()

ui.run()
```

results in

```
    raise RuntimeError('cannot find parent view, view stack is empty')
RuntimeError: cannot find parent view, view stack is empty
```

I have just recently migrated a great chunks of previously justpy code to the above inheritance approach including additional element/group classes (such as header, layout, pagecontainer, ...) and am now seeking ways to make it work, again.

@falkoschindler: could you please guide me here, again? Thanks!

This issue is possibly related to #85.

**Comment by rodja** [comment]
We now follow this pattern when we want to wrap a page into their own class:

```python
class MainPage():
    def __init__(self):
        super().__init__()

        @ui.page('/', dark=True)
        def page():
            ui.label('Hi!')

MainPage()
```

If you do not need or want to register the page within the constructor of a class, you can simplify the above to 

```python
def create_main_page():
    @ui.page('/', dark=True)
    def page():
        ui.label('Hi!')

create_main_page()
```

Or, as FastAPI is doing it: just create a new file `main_page.py` containing:

```python
from nicegui import ui

@ui.page('/', dark=True)
def page():
    ui.label('Hi!')
```

and then in your `main.py` simply do:

```python
from nicegui import ui
import main_page

ui.run()
```

As you can see, there is no need for inheritance anymore.

**Comment by falkoschindler** [comment]
Another pattern might be to define the content in a class and do the routing in main.py:

```python
class MainPage:
    def __init__(self):
        ui.label('Hi!')
```

```python
@ui.page('/')
def main():
    MainPage()
```

---

Or define the page content in a function:

```python
def main_page():
    ui.label('Hi!')
```

```python
@ui.page('/')
def main():
    main_page()
```

**Comment by rodja** [comment]
Thanks for your input, @hroemer. This helped me better to understand your requirement with encapsulating common behaviour and styling like header/footer etc in a custom "base" page.
We just released [v0.9.2](https://github.com/zauberzeug/nicegui/releases/tag/v0.9.2). Now you can derive from `@ui.page` to create your own custom page decorator. I created the [customisation example](https://github.com/zauberzeug/nicegui/tree/v0.9.2/examples/customisation) to show this in action. Does that work for you?

The other requirement I'm guessing you have is to create new "Quasar" elements. Is that correct? Maybe you could open a separate issue for this?

**Comment by rodja** [comment]
@hroemer to 1:

I got the naming wrong. In [0.9.3](https://github.com/zauberzeug/nicegui/releases/tag/v0.9.3) I used your suggestions header/footer. In the example I used the header to [open up a context manager to align all content](https://github.com/zauberzeug/nicegui/blob/f417d7a64b43f00e52dd843445a8a9d8b9b366b9/examples/customisation/custom.py#L16-L17).

Also in [0.9.3](https://github.com/zauberzeug/nicegui/releases/tag/v0.9.3) I made the page available in header and footer via `self.page`. See the [customized_page test](https://github.com/zauberzeug/nicegui/blob/f417d7a64b43f00e52dd843445a8a9d8b9b366b9/tests/test_pages.py#L136-L158).

But your need to access the data property got me thinking: for what purpose do you use the justpy binding in the first place? I'll need to dig deeper into your example....

**Comment by me21** [comment]
I'm sorry to piggyback on the topic, but is it currently possible to rewrite LoggedInPage and LoggedInPrivatePage from https://github.com/zauberzeug/nicegui/issues/81 using decorator syntax?

**Comment by rodja** [comment]
@me21 I just released [0.9.4](https://github.com/zauberzeug/nicegui/releases/tag/v0.9.4) which adds the possibility to receive the request object for page building. With this it was fairly easy to create an [authentication example](https://github.com/zauberzeug/nicegui/releases/tag/v0.9.4).

**Comment by rodja** [comment]
@hroemer I created a Q&A discussion for the wrapping of Quasar elements: #99
